GiniVision/Classes/GiniQRCodeDocument.swift

```diff
@@ -0,0 +1,162 @@
+//
+//  GiniQRCodeDocument.swift
+//  Bolts
+//
+//  Created by Enrique del Pozo Gómez on 12/5/17.
+//
+
+import Foundation
+
+@objc final public class GiniQRCodeDocument: NSObject, GiniVisionDocument {
+    public var type: GiniVisionDocumentType = .qrcode
+    public var data: Data
+    public lazy var previewImage: UIImage? = {
+        return self.generateQRCodeImage(from: self.data)
+    }()
+    public var isReviewable: Bool = false
+    public var isImported: Bool = false
+    public lazy var extractedParameters: [String: String] = self.extractParameters(from: self.scannedString)
+    
+    fileprivate let scannedString: String
+    fileprivate let epc06912LinesCount = 12
+    fileprivate lazy var qrCodeFormat: QRCodesFormat? = {
+        if self.scannedString.starts(with: "bank://") {
+            return .bezahlcode
+        } else {
+            let lines = self.scannedString.splitlines
+            if lines.count > 9 &&
+                (lines[1] == "001" || lines[1] == "002") &&
+                (lines[2] == "1" || lines[2] == "2") {
+                return .epc06912
+            }
+            return nil
+        }
+    }()
+    
+    fileprivate enum QRCodesFormat {
+        case epc06912
+        case bezahlcode
+    }
+    
+    public init(scannedString: String) {
+        self.data = scannedString.data(using: String.Encoding.isoLatin1) ?? Data(count: 0)
+        self.scannedString = scannedString
+        super.init()
+    }
+    
+    public func checkType() throws {
+        if self.qrCodeFormat == nil || self.extractedParameters.isEmpty {
+            throw DocumentValidationError.qrCodeFormatNotValid
+        }
+    }
+    
+    fileprivate func generateQRCodeImage(from data: Data) -> UIImage? {
```

Comment:
Why is there an image generation method in a model class?
Transitively, you have a dependency yo UIKit is a model class (UIImage).

Reply:
I was thinking about this since I wrote it, I wasn't sure about it. Yes, definitely this should me move to other class.